src/libraries/System.Net.Http/ref/System.Net.Http.cs

```diff
@@ -299,6 +301,23 @@ public sealed class HttpRequestOptions : System.Collections.Generic.IDictionary<
         public void Set<TValue>(HttpRequestOptionsKey<TValue> key, TValue value) { throw null; }
     }
 
+    [SupportedOSPlatform("windows")]
+    [SupportedOSPlatform("linux")]
+    [SupportedOSPlatform("macos")]
+    [UnsupportedOSPlatform("browser")]
+
+    public sealed class Http3WebtransportSession : System.IAsyncDisposable, IDisposable
+    {
+        public long Id { get { throw null; } }
+        internal Http3WebtransportSession() { }
+        public static System.Threading.Tasks.ValueTask<Http3WebtransportSession?> ConnectAsync(Uri uri, HttpMessageInvoker invoker, System.Threading.CancellationToken cancellationToken) { throw null; }
+        public bool TryGetIncomingWebtransportStreamFromServer(out System.Net.Quic.QuicStream? quicStream) { throw null; }
+        public System.Threading.Tasks.ValueTask<System.Net.Quic.QuicStream?> GetIncomingWebtransportStreamFromServerAsync(CancellationToken cancellationToken = default) { throw null; }
+        public System.Threading.Tasks.ValueTask<System.Net.Quic.QuicStream?> OpenWebtransportStreamAsync(System.Net.Quic.QuicStreamType type, CancellationToken cancellationToken = default) { throw null; }
+        public System.Threading.Tasks.ValueTask DisposeAsync() { throw null; }
```

Comment:
I think you can have only DisposeAsync, you don't have to have both

Reply:
The functions which are closing the quic connection are not async, so I also created a simple Dispose when I close because of the connection being closed. Should I make the simple Dispose private?

Reply:
QuicConnection actually *only* have DisposeAsync, but I think you meant QuicStream instead? It has both Dispose and DisposeAsync, and they are doing the same thing (more specifically, stream's Dispose is calling DisposeAsync inside). So you can only use DisposeAsync.

It is also no problem to call a sync method from async code. It might be a problem to do another way around (calling async method from sync code, known as "sync-over-async") and it is not advisable unless you 100% sure what you're doing and there's no other way.

Reply:
I meant the `CheckForShutdown` function inside H3Connection which handles the task for closing the quic connection, it is a sync function and I cannot use Dispose async inside it. Maybe I should place the disposal of the manager inside another part of the code so i can get rid of dispose?

Reply:
I see, you were talking about disposal chain connection->manager->session. Not only it is a sync function, but it also is being called from inside the lock, so you cannot use async there. In that case we might want to either only keep Dispose, or to keep both Dispose and DisposeAsync.

Couple of thoughts:

1) we need to be careful with the locks, as soon as there are 2 locks in the picture (one in H3Connection and other inside the manager that we've added) we need to assess probability of a deadlock.

2) Related to discussion of active requests, we need to think whether the dispose in a *graceful* scenario is actually needed. If session streams would count towards active requests, that means we wouldn't get past these lines https://github.com/iuliaco/runtime/blob/54858d80d3c7b640386ba1af9f56118aac60014b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L129-L132 until all the sessions are stopped and removed from the manager. Which is good actually. In an abortive scenario, the connection itself is aborted in advance, which should also trigger the disposal of sessions. So if everything is aligned properly, we might not need this disposal at all. Am I making sense? :D

Reply:
@CarnaViire I made the removal of the sessions from the `_active requests` when I remove session from the dictionary since that means that the session will no longer be active. Let me know what you think about my modifications